### ISSUE TITLE
Not getting good results

### ISSUE THREAD

**Issue by julianpalladino** [simontomaskarlsson/CycleGAN-Keras#6]
Hi there! Thanks for this implementation in Keras, it looks very organized and explanatory.
However, I could not achieve good results after 200 epochs with two of the the [original datasets](https://github.com/junyanz/CycleGAN#datasets).

I tried with the photo2vangogh dataset, obtaining:
![epoch198_sample0](https://user-images.githubusercontent.com/8797947/58808630-c5e0cc80-85f0-11e9-9265-b468b8db988a.png)

And with horses2zebras I obtained:
![epoch221_sample1](https://user-images.githubusercontent.com/8797947/58808582-aba6ee80-85f0-11e9-85aa-2bedc63bca90.png)

Did you modify the original architecture in any way in order to work better for your dataset of night/day streets?

Is there any advice you can give me to get better results with these other datasets?

Thanks in advance!

**Comment by simontomaskarlsson** [comment]
Hi @julianpalladino ,

Since Keras and tensorflow are different by design i can't say that the models are identical. I did however spend a lot of time trying to implement this model as close to the original as possible. After I could attain good results on the original training sets I, for research purposes, implemented more options to the model that you can enable. 

The default model options gives you the original model architecture and should provide a good result, can you verify that you are using these?

As advice I can say that I do recognize the problem you are having, both the pattern and the unsatisfying color in the synthetic images. The model seems to converge to something it can easily reconstruct to an image close to the original. You could try the model without the latest update to the model (https://github.com/simontomaskarlsson/CycleGAN-Keras/commit/1572dd0b5fab3a2de52919d11eca2b1a787c53e5) since it worked for me before this.
You can also try and change the normalisation method to fix the pattern issue.

**Comment by julianpalladino** [comment]
Hi @simontomaskarlsson, thank you for your quick answer!
I just tried the model without the latest update to the model, but I'm getting the same results.
As for the settings, I am using TensorFlow 2.0, so I had to modify the libraries imported in the project ("from tensorflow.keras import..." instead of "from keras import..."), but besides that I kept all the settings default and didn't modify anything.


> The model seems to converge to something it can easily reconstruct to an image close to the original.

It certainly looks like that! The reconstruction is perfect but the generation is bad.


Other examples from this run to ilustrate this:
![epoch74_sample0](https://user-images.githubusercontent.com/8797947/58880573-9c867600-86ae-11e9-90d5-e66269a59132.png)

![epoch83_sample0](https://user-images.githubusercontent.com/8797947/58880532-7a8cf380-86ae-11e9-8f25-d08e59ec125b.png)
It looks like it just inverts the colors, and then reconstructs perfectly...

Do you have any other idea I could try? I was thinking about giving more weight to the generator loss, so the model tries to prioritize it over the reconstruction loss. But I don't understand yet how the losses are connected between the networks. Is it possible to do this by modifying one of the following?
```
self.lambda_1 = 10.0  # Cyclic loss weight A_2_B
self.lambda_2 = 10.0  # Cyclic loss weight B_2_A
self.lambda_D = 1.0  # Weight for loss from discriminator guess on synthetic images
```

Thanks again!

**Comment by simontomaskarlsson** [comment]
Modifying the loss weights is something worth exploring I think. Also while doing this, look at the losses that are printed out. You should have all the losses stored in a csv-file after each training, see the function 'writeLossDataToFile' in model.py for details. If you want you can use 'plotCSVfile.py' for visualization.

This image explains the weight update of the generators and discriminators.
![image](https://user-images.githubusercontent.com/17566958/58938642-8ee7f380-8775-11e9-926e-75ebf412b4ff.png)

Especially, " The generator weight update depends on the cyclic loss and the output from discriminator" and "Discriminator weights are updated depending on the output from real and
synthetic images"

Since the reconstruction is perfect you might want to lower the lambda_1 and lambda_2 to get what you are looking for. But also consider having a look at the losses throughout the training since that might give you some clues. My guess is that the discriminators stop updating their weights after a number of epochs, probably because they always estimate the both the synthetic and real images correcly.

**Comment by julianpalladino** [comment]
Wow, thanks for the complete explanation!

First I tried:
```
self.lambda_1 = 10.0 # Cyclic loss weight A_2_B
self.lambda_2 = 10.0  # Cyclic loss weight B_2_A
self.lambda_D = 10.0   # Weight for loss from discriminator guess on synthetic images
```
obtaining:
![epoch147_sample1](https://user-images.githubusercontent.com/8797947/59035441-aee5e880-8843-11e9-93ce-75bf6b3683c8.png)

And after I tried:
```
self.lambda_1 = 2.0 # Cyclic loss weight A_2_B
self.lambda_2 = 2.0  # Cyclic loss weight B_2_A
self.lambda_D = 1.0    # Weight for loss from discriminator guess on synthetic images
```
obtaining:
![epoch111_sample1](https://user-images.githubusercontent.com/8797947/59035497-cc1ab700-8843-11e9-9781-8d6030477b15.png)

Which looks pretty great compared to the previous results. However, some images still have some problems, such as:
![epoch158_sample1](https://user-images.githubusercontent.com/8797947/59035567-f10f2a00-8843-11e9-8932-e871c266c325.png)

So I'll keep trying with different weights until I get a better result.


As for the weights, I'm not sure I understood which one is lambda1, lambda2 and lambdaD, because in the original paper the loss formula looks like this:

![Screenshot from 2019-06-06 10-28-15](https://user-images.githubusercontent.com/8797947/59036621-0d13cb00-8846-11e9-9211-fbe6fe48aee1.png)


It seems to have one loss for each GAN and then another loss for Cycle Consistency. But you have the opposite, one loss for each Cycle consistency and one for both GANs (right?)

**Comment by julianpalladino** [comment]
Ah, as for the 'plotCSVfile.py', I obtained:
![Screenshot from 2019-06-06 10-42-58](https://user-images.githubusercontent.com/8797947/59037640-0b4b0700-8848-11e9-8fda-801770305db9.png)

Which I think it's telling me the reconstruction is too good and generation too bad, and I should keep lowering lambda1 and lambda2 to fix it

**Comment by simontomaskarlsson** [comment]
Nice, looks like you are getting somewhere. If you have time, also try batch normalization instead of instance normalization. I don't have a good explanation to why it would help but I'm hoping it could get rid of the pattern artifact you have in some images. Also, `use_identity_learning = true` could help you with the problem the model seems to have with the color of the sky.

As for the losses they are exactly the same when you are using the default hyper parameter values, i.e. 
`self.lambda_1 = 10.0  # Cyclic loss weight A_2_B
self.lambda_2 = 10.0  # Cyclic loss weight B_2_A
self.lambda_D = 1.0  # Weight for loss from discriminator guess on synthetic images`

In the original implementation the values on lamda_1 and lambda_2 are always the same, which means they only need one variable, and the value on lambda_D is always equal to 1.0, which means they do not need a variable. What I did was adding variables for flexibility and using them in my research.

As for the plots I agree with you. My main concern is however the thing I expected, "that the discriminators stop updating their weights after a number of epochs, probably because they always estimate the both the synthetic and real images correcly" which you can see on the zero discriminator loss. Have that in mind also and keep doing what you are doing! Also, take a look inside 'plotCSVfile.py' and modify it if you want to plot something specific.
